feat: request new tokens before they expire

ForbesLindesay opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/node-gtoken/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #230 🦕

ForbesLindesay committed:
feat: request new tokens before they expire

ForbesLindesay commented:
@googlebot I signed it!

ForbesLindesay committed:
test: add test cases for expiresSoon

ForbesLindesay committed:
fix: match behaviour from google-auth-library-nodejs

[closes #230]

bcoe commented on test/index.ts:
```diff
@@ -336,6 +361,18 @@ describe('.getToken()', () => {
+    gtoken.expiresAt = new Date().getTime() + 4 * 60 * 1000;
+    const fakeToken = 'abc123';
+    const scope = createGetTokenMock(200, {access_token: fakeToken});
+    const token = await gtoken.getToken({forceRefresh: true});
```
If possible, let's get rid of some of the warnings. Can you call `scope.done()` to ensure that the mock finished, as an example?

bcoe commented on test/index.ts:
```diff
@@ -138,6 +138,31 @@ describe('.hasExpired()', () => {
+  });
+
+  it('should detect expiring tokens', () => {
+    const gtoken = new GoogleToken();
```
For ease of migrating this behavior from the auth library to the token library, I think we should make the functionality disabled by default.

bcoe reviewed (changes requested):
This seems like a solid addition to the library, I think we might want to make the eagerRefresh behavior opt in though (with a test that makes sure it's not applied by default).

This will make it easier to move the auth library over to this version of gtoken.

ForbesLindesay committed:
Make eagerRefreshThresholdMillis default to 0

This keeps behaviour consistent with what it was before by default. I've restored the original tests, so this just adds new tests for the new functionality.

ForbesLindesay committed:
docs: update comment to match new default

ForbesLindesay committed:
fix: cleanup some accidental changes to tests

ForbesLindesay committed:
Merge branch 'master' into patch-1

ForbesLindesay commented:
I've restored the original default behaviour of 0 ms for eager refresh. I've tried to keep my change fairly minimal to make it easy to merge, but I did add the `scope` assertions as requested.

bcoe commented on test/index.ts:
```diff
@@ -138,6 +138,33 @@ describe('.hasExpired()', () => {
+describe('.isTokenExpiring()', () => {
+  it('should exist', () => {
+    const gtoken = new GoogleToken();
+    assert.strictEqual(typeof gtoken.isTokenExpiring, 'function');
```
I think we should add a test for the default behavior of `eagerRefreshThresholdMillis = 0`, which I believe should always indicate `isTokenExpiring()` of `false?

ForbesLindesay replied:
I've added a test for this

bcoe reviewed (commented):
👍 this is looking good to me; I might be missing it, but I think one test for the default behavior of not refreshing would be good.

Once we release this, we can look at changing the behavior of `google-auth-library`. Out of curiosity, are you using gtoken directly, or using it through the auth library?

ForbesLindesay commented:
> I might be missing it, but I think one test for the default behavior of not refreshing would be good.

The original tests already cover this case. The library always refreshed expired tokens.

See "should return cached token if not expired"

ForbesLindesay committed:
Add yet more tests

ForbesLindesay commented:
We're using it directly. The layers on top just added masses of complexity that were extremely difficult to dig through when all we needed was a token to use in a fetch request.

ForbesLindesay committed:
Merge branch 'master' into patch-1

bcoe commented:
@ForbesLindesay not quite sure why CI/CD is failing after the last update. If we can get things green, let's land this.

ForbesLindesay committed:
fix: update method after it was renamed

Commit:
Merge branch 'master' into patch-1

bcoe commented:
@ForbesLindesay thank you for the contribution 👍 

This is now out the door in `5.2.0`, let me know if it's working well for you?
